### DISCUSSION TITLE
Scaling DB containers horizontally?

### DISCUSSION THREAD

**Discussion by DAB0mB** [appwrite/appwrite#3621] [Q&A] [answered]
I've been looking at the docs, specifically at the [scaling](https://appwrite.io/docs/production#scaling) section. It's understandable how to scale horizontally for stateless containers, you just create replicas, but what about stateful DB containers? i.e.MariaDB, Redis, and InfluxDB containers. I didn't find any reference to AppWrite knowing how to work with DB replicas or sharding, does this mean that you can only scale these vertically? Is there any hard limit I should be concerned of? Does anybody have experience with that or ran some stress tests? A client of mine has a potential of 10mil active monthly users and I wanted to know if it's realistic to expect AppWrite to handle it.

**Comment by stnguyen90** [comment, marked as answer]
@DAB0mB, you would scale MariaDB just like you would in any other environment. Maybe this will help: https://mariadb.com/kb/en/replication-cluster-multi-master/. You can also use a managed mariadb provider if you don't want to manage the scaling yourself.

**Reply by DAB0mB** [reply]
If I understand correctly, you suggest to use multiple MariaDB containers that share the same DB connection variables? Given that I can use a managed service like AWS RDS.

**Reply by ammarfaris** [reply]
As I understand, to take advantage of read replicas of MariaDB, we need to configure Appwrite application to direct read-only database queries to the replicas, while all write operations continue to go to the primary database.


Is this supported natively? @stnguyen90 

_Chat GPT gave me this answer:_

> 
> **Appwrite’s Database Layer**
> Appwrite itself does not have built-in, automatic support for utilizing read replicas directly.
> Appwrite expects a central database endpoint (MySQL/MariaDB/Postgres) and handles all reads/writes through this connection.
> Appwrite does not natively route read queries to replicas.
